packages/phone-number-privacy/combiner/src/combiner/combiner.service.ts

```diff
@@ -0,0 +1,309 @@
+
+export interface SignerResponseWithStatus {
+  url: string
+  res: SignerResponse
+  status: number
+}
+
+export interface SignerService {
+  url: string
+  fallbackUrl?: string
+}
+
+export abstract class CombinerService implements ICombinerService {
+  protected failedSigners: Set<string>
+  protected errorCodes: Map<number, number>
+  protected timedOut: boolean
+  protected signers: SignerService[]
+  protected timeoutMs: number
+  protected threshold: number
+  protected enabled: boolean
+  protected logger: Logger
+  protected abstract endpoint: CombinerEndpoint
+  protected abstract signerEndpoint: SignerEndpoint
+  protected abstract responses: SignerResponseWithStatus[]
+
+  public constructor(config: OdisConfig, protected io: IInputService) {
+    this.logger = rootLogger()
+    this.failedSigners = new Set<string>()
+    this.errorCodes = new Map<number, number>()
+    this.timedOut = false
+    this.signers = JSON.parse(config.odisServices.signers)
+    this.timeoutMs = config.odisServices.timeoutMilliSeconds
+    this.threshold = config.keys.threshold
+    this.enabled = config.enabled
+  }
+
+  public async handleDistributedRequest(
+    request: Request<{}, {}, DistributedRequest>,
+    response: Response
+  ) {
+    this.logger = response.locals.logger
+    try {
+      if (!(await this.inputCheck(request, response))) {
+        return
+      }
+      await this.forwardToSigners(request)
+      await this.combineSignerResponses(request, response)
+    } catch (err) {
+      this.logger.error(`Unknown error in handleDistributedRequest for ${this.endpoint}`)
+      this.logger.error(err)
+      this.sendFailureResponse(response, ErrorMessage.UNKNOWN_ERROR, 500)
+    }
+  }
+
+  protected async inputCheck(
+    request: Request<{}, {}, DistributedRequest>,
+    response: Response
+  ): Promise<boolean> {
+    if (!this.enabled) {
+      this.sendFailureResponse(response, WarningMessage.API_UNAVAILABLE, 501)
+      return false
+    }
+    if (!this.io.validate(request, this.logger)) {
+      this.sendFailureResponse(response, WarningMessage.INVALID_INPUT, 400)
+      return false
+    }
+    if (!(await this.io.authenticate(request, this.logger))) {
+      this.sendFailureResponse(response, WarningMessage.UNAUTHENTICATED_USER, 401)
+      return false
+    }
+    return true
+  }
+
+  protected async forwardToSigners(request: Request<{}, {}, DistributedRequest>) {
+    // TODO(Alec): Factor out metering code
+    const obs = new PerformanceObserver((list) => {
+      const entry = list.getEntries()[0]
+      this.logger.info({ latency: entry, signer: entry!.name }, 'Signer response latency measured')
+    })
+    obs.observe({ entryTypes: ['measure'], buffered: true })
+
+    const controller = new AbortController()
+    const timeout = setTimeout(() => {
+      this.timedOut = true
+      controller.abort()
+    }, this.timeoutMs)
```

Comment:
IIUC, `AbortController` has its own timeout logic. If possible, we should use that instead and possibly remove the `timedOut` variable https://dom.spec.whatwg.org/#dom-abortsignal-timeout

Reply:
I looked into this a little bit, and couldn't find much on this outside of that link. There isn't any timeout logic on the AbortSignal type I'm currently using, and most advice online seems to be to use the pattern we're currently using. Unless you feel strongly, I'd propose that we keep this as is for the time being